py/kubeflow/testing/util.py

```diff
@@ -47,26 +47,32 @@ def run(command, cwd=None, env=None, dryrun=False):
     logging.info("Running: Environment:\n%s", "\n".join(lines))
 
   log_file = None
-  try:
-    if dryrun:
-      command_str = ("Dryrun: Command:\n{0}\nCWD:\n{1}\n"
-                     "Environment:\n{2}").format(" ".join(command), cwd, env)
-      logging.info(command_str)
-
-    # We write stderr/stdout to a file and then read it and process it.
-    # We do this because if just inherit the handles from the parent the
-    # subprocess output doesn't show up in Airflow. This might be because
-    # we had multiple levels of processes invoking python processes.
-    with tempfile.NamedTemporaryFile(prefix="tmpRunLogs", delete=False,
-                                     mode="w") as hf:
-      log_file = hf.name
-      subprocess.check_call(command, cwd=cwd, env=env,
-                            stdout=hf,
-                            stderr=hf)
-  finally:
-    with open(log_file, "r") as hf:
-      output = hf.read()
-    logging.info("Subprocess output:\n%s", output)
+  
+  if dryrun:
+    command_str = ("Dryrun: Command:\n{0}\nCWD:\n{1}\n"
+                   "Environment:\n{2}").format(" ".join(command), cwd, env)
+    logging.info(command_str)
+
+  process = subprocess.Popen(
+    command, cwd=cwd, env=env, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+  logging.info("Subprocess output:\n")
+  output = []
+  while process.poll() is None:
+    process.stdout.flush()
+    for line in iter(process.stdout.readline, ''):
+      output.append(line.strip())
+      logging.info(line.strip())
+
+  process.stdout.flush()
+  for line in iter(process.stdout.readline, ''):
+    output.append(line.strip())
+    logging.info(line.strip())
+
+  if process.returncode != 0:
+    raise subprocess.CalledProcessError(process.returncode,
+                                        "cmd: {0} exited with code {1}".format(
+                                        " ".join(command), process.returncode), "\n".join(output))
```

Comment:
output should be printed before, do we need to print again?

Reply:
Output is being included in the exception. This makes it consistent with the exception that would be raised if we used subprocess.check_call